pkg/build/admission/jenkinsbootstrapper/admission.go

```diff
@@ -0,0 +1,162 @@
+)
+
+func init() {
+	admission.RegisterPlugin("JenkinsBootstrapper", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
+		return NewJenkingsBootstrapper(c.Core()), nil
+	})
+}
+
+type jenkingsBootstrapper struct {
+	*admission.Handler
+
+	privilegedRESTClientConfig restclient.Config
+	serviceClient              coreclient.ServicesGetter
+	openshiftClient            client.Interface
+
+	jenkinsConfig configapi.JenkinsPipelineConfig
+}
+
+// NewJenkingsBootstrapper returns an admission plugin that will create required jenkins resources as the user if they are needed.
+func NewJenkingsBootstrapper(serviceClient coreclient.ServicesGetter) admission.Interface {
+	return &jenkingsBootstrapper{
+		Handler:       admission.NewHandler(admission.Create),
+		serviceClient: serviceClient,
+	}
+}
+
+func (a *jenkingsBootstrapper) Admit(attributes admission.Attributes) error {
+	if a.jenkinsConfig.Enabled != nil && !*a.jenkinsConfig.Enabled {
+		return nil
+	}
+	if len(attributes.GetSubresource()) != 0 {
+		return nil
+	}
+	if attributes.GetResource().GroupResource() != buildapi.Resource("buildconfigs") && attributes.GetResource().GroupResource() != buildapi.Resource("builds") {
+		return nil
+	}
+	if !needsJenkinsTemplate(attributes.GetObject()) {
+		return nil
+	}
+
+	namespace := attributes.GetNamespace()
+	glog.V(4).Infof("Handling %v %s/%s", attributes.GetResource(), namespace, attributes.GetName())
+
+	svcName := a.jenkinsConfig.ServiceName
+	if len(svcName) == 0 {
+		return nil
+	}
+
+	// TODO pull this from a cache.
+	// if its not a not-found error (no error or anything else) ,just return ok an take no action.  hopefully it will work.
+	if _, err := a.serviceClient.Services(namespace).Get(svcName); !kapierrors.IsNotFound(err) {
+		if err == nil {
+			glog.V(4).Infof("The Jenkins Pipeline service %q already exists in project %q", svcName, namespace)
+		} else {
+			glog.V(2).Infof("error getting Jenkins Pipeline service %q already exists in project %q: %v", svcName, namespace, err)
+		}
+		return nil
+	}
+
+	glog.V(3).Infof("Adding new Jenkins service %q to the project %q", svcName, namespace)
+	jenkinsTemplate := jenkinscontroller.NewPipelineTemplate(namespace, a.jenkinsConfig, a.openshiftClient)
+	objects, errs := jenkinsTemplate.Process()
+	if len(errs) > 0 {
+		return kutilerrors.NewAggregate(errs)
+	}
+	if !jenkinsTemplate.HasJenkinsService(objects) {
+		return fmt.Errorf("template %s/%s does not contain required service %q", a.jenkinsConfig.TemplateNamespace, a.jenkinsConfig.TemplateName, a.jenkinsConfig.ServiceName)
+	}
+
+	impersonatingConfig := a.privilegedRESTClientConfig
+	oldWrapTransport := impersonatingConfig.WrapTransport
+	impersonatingConfig.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+		return authenticationclient.NewImpersonatingRoundTripper(attributes.GetUserInfo(), oldWrapTransport(rt))
+	}
+
+	var bulkErr error
+
+	bulk := &cmd.Bulk{
+		Mapper: &resource.Mapper{
```

Comment:
would it make sense to move this to a common utility func? like impersonatingMapper(blah)